## Accessing and Manipulating Kaguya Spectral Profiler Data

First, it is necessary to import the `libpyhat` module.  This notebook also imports a helper function `get_path` that makes working with the sample data shipped with `libpyhat` easier.

In [6]:
import libpyhat as phat
from libpyhat.examples import get_path
import plio.io.io_spectral_profiler as isp

['/Users/tgabriel/PyHAT/notebooks', '/Users/tgabriel/anaconda3/envs/pyhat/lib/python39.zip', '/Users/tgabriel/anaconda3/envs/pyhat/lib/python3.9', '/Users/tgabriel/anaconda3/envs/pyhat/lib/python3.9/lib-dynload', '', '/Users/tgabriel/anaconda3/envs/pyhat/lib/python3.9/site-packages', '/Users/tgabriel/anaconda3/envs/pyhat/lib/python3.9/site-packages/IPython/extensions', '/Users/tgabriel/.ipython', 'PyHAT/libpyhat', 'PyHAT/libpyhat', '/Users/tgabriel/PyHAT/libpyhat', '/Users/tgabriel/PyHAT/libpyhat/']
['/Users/tgabriel/PyHAT/notebooks', '/Users/tgabriel/anaconda3/envs/pyhat/lib/python39.zip', '/Users/tgabriel/anaconda3/envs/pyhat/lib/python3.9', '/Users/tgabriel/anaconda3/envs/pyhat/lib/python3.9/lib-dynload', '', '/Users/tgabriel/anaconda3/envs/pyhat/lib/python3.9/site-packages', '/Users/tgabriel/anaconda3/envs/pyhat/lib/python3.9/site-packages/IPython/extensions', '/Users/tgabriel/.ipython', 'PyHAT/libpyhat', 'PyHAT/libpyhat', '/Users/tgabriel/PyHAT/libpyhat', '/Users/tgabriel/PyHAT/li

To open a spectral profiler 'image', we use the `plio.io.io_spectral_profiler.Spectral_Profiler` call.  If example data is not going to be used, the `get_path(<my_file_path>)` can be replaced with `<my_file_path>`, since our helper function does not know where your data is being stored. 

In [7]:
profile = isp.Spectral_Profiler(get_path('SP_2C_02_02358_S138_E3586.spc'))

You may want to learn a bit more about the spectra you're about to analyze/process, so let's read some of the header information.

In [155]:
profile.label

PVLModule([
  ('PDS_VERSION_ID', 'PDS3')
  ('RECORD_TYPE', 'UNDEFINED')
  ('FILE_NAME', 'SP_2C_02_02358_S138_E3586.spc')
  ('PRODUCT_ID', 'SP_2C_02_02358_S138_E3586')
  ('DATA_FORMAT', 'PDS')
  ('^ANCILLARY_AND_SUPPLEMENT_DATA', Units(value=24737, units='BYTES'))
  ('^SP_SPECTRUM_WAV', Units(value=31045, units='BYTES'))
  ('^SP_SPECTRUM_RAW', Units(value=31637, units='BYTES'))
  ('^SP_SPECTRUM_REF2', Units(value=54133, units='BYTES'))
  ('^SP_SPECTRUM_RAD', Units(value=76629, units='BYTES'))
  ('^SP_SPECTRUM_REF1', Units(value=99125, units='BYTES'))
  ('^SP_SPECTRUM_QA', Units(value=121621, units='BYTES'))
  ('^L2D_RESULT_ARRAY', Units(value=144117, units='BYTES'))
  ('SOFTWARE_NAME', 'RGC_SP')
  ('SOFTWARE_VERSION', '2.10.3')
  ('PROCESS_VERSION_ID', 'L2C')
  ('PRODUCT_CREATION_TIME',
   datetime.datetime(2012, 3, 28, 18, 21, 14, tzinfo=<UTC>))
  ('PROGRAM_START_TIME', datetime.datetime(2012, 3, 28, 18, 15, 41, tzinfo=<UTC>))
  ('PRODUCER_ID', 'LISM')
  ('PRODUCT_SET_ID', 'SP_Level2C'

There are other bits of information that held in the attributes of the `profile` object, such as the number of obseravations.

In [175]:
profile.nspectra

38

Now lets access the spectra itself.

In [176]:
s = profile.spectra

The `s` object is based on a pandas data frame. Therefore, anything that you might normally do with a pandas data frame, can be applied to the object. In this notebook we demo a few of the possible operations that Pandas provides. Note that the parent object of `s` (`profile`) has other functionalities you can access such as `profile.wavelength`.

## Viewing the data

To see the first or last rows in the `Spectra` object, one can use `head` or `tail`, respectively.

In [177]:
#Show the first 5 lines of the pandas frame
s.head(5)

0                                     1                          \
          RAW    REF1    REF2     QA    RAD     RAW    REF1    REF2     QA   
512.6  5123.0  0.0402  0.0397  288.0  27.94  5164.0  0.0414  0.0409  288.0   
518.4  5887.0  0.0487  0.0482  288.0  30.99  5955.0  0.0502  0.0497  288.0   
524.7  6375.0  0.0497  0.0492  288.0  33.66  6461.0  0.0513  0.0507  288.0   
530.4  6806.0  0.0520  0.0515  288.0  35.97  6905.0  0.0537  0.0531  288.0   
536.5  7494.0  0.0532  0.0527  288.0  36.76  7614.0  0.0549  0.0544  288.0   

              ...      36                                    37          \
         RAD  ...     RAW    REF1    REF2     QA    RAD     RAW    REF1   
512.6  28.77  ...  5120.0  0.0402  0.0398  288.0  27.72  5046.0  0.0387   
518.4  31.96  ...  5881.0  0.0488  0.0483  288.0  30.78  5774.0  0.0468   
524.7  34.72  ...  6371.0  0.0498  0.0493  288.0  33.42  6233.0  0.0477   
530.4  37.09  ...  6799.0  0.0521  0.0516  288.0  35.70  6641.0  0.0500   
536.5  37.93  ...  7486.0  0.0533  0.0527  288.0  36.46  7289.0  0.0511   

                             
         REF2     QA    RAD  
512.6  0.0383  288.0  26.66  
518.4  0.0464  288.0  29.54  
524.7  0.0473  288.0  32.06  
530.4  0.0495  288.0  34.24  
536.5  0.0506  288.0  34.96  

[5 rows x 190 columns]

In [178]:
#Show the last 5 lines of the pandas frame
s.tail(5)

0                                     1                          \
            RAW    REF1    REF2     QA   RAD      RAW    REF1    REF2     QA   
2556.0   6806.0  0.0000  0.0000  296.0  0.00   6815.0  0.0000  0.0000  296.0   
2564.0  11534.0  0.0000  0.0000  296.0  0.00  11526.0  0.0000  0.0000  296.0   
2572.0   7317.0  0.0000  0.0000  296.0  0.00   7322.0  0.0000  0.0000  296.0   
2579.9  11412.0  0.0213  0.0212  288.0  0.34  11410.0  0.0125  0.0125  288.0   
2587.9   6100.0  0.0000  0.0000  296.0  0.00   6095.0  0.0000  0.0000  296.0   

             ...       36                                   37          \
        RAD  ...      RAW    REF1    REF2     QA  RAD      RAW    REF1   
2556.0  0.0  ...   6809.0  0.0000  0.0000  296.0  0.0   6812.0  0.0000   
2564.0  0.0  ...  11521.0  0.0000  0.0000  296.0  0.0  11522.0  0.0000   
2572.0  0.0  ...   7330.0  0.0000  0.0000  296.0  0.0   7326.0  0.0000   
2579.9  0.2  ...  11410.0  0.0126  0.0126  288.0  0.2  11416.0  0.0385   
2587.9  0.0  ...   6095.0  0.0000  0.0000  296.0  0.0   6092.0  0.0000   

                             
          REF2     QA   RAD  
2556.0  0.0000  296.0  0.00  
2564.0  0.0000  296.0  0.00  
2572.0  0.0000  296.0  0.00  
2579.9  0.0383  288.0  0.61  
2587.9  0.0000  296.0  0.00  

[5 rows x 190 columns]

## Accessing individual observations

Above, the data is segmented by an index in the first row, from observation 0 to 37. The first column represents the wavelength, which ranges from from ~500 to ~2600. The five columns below each observation number correspond to: (1) the raw observed spectra (RAW), (2) a Lunar Mare-corrected continuum (REF1), (3) a Lunar Highlands-corrected continuum (REF2), and (4) .a team provided quality assurance row (QA), (2) the raw observered spectra (RAW), (3) a mare corrected continuum (REF1), and (4) the radiance offset for the spectra. If we want to work with a single observation, say observation number 0, we can just use the index like so:

In [179]:
#Show the first observation, across all wavelengths
s[0]

RAW    REF1    REF2     QA    RAD
512.6    5123.0  0.0402  0.0397  288.0  27.94
518.4    5887.0  0.0487  0.0482  288.0  30.99
524.7    6375.0  0.0497  0.0492  288.0  33.66
530.4    6806.0  0.0520  0.0515  288.0  35.97
536.5    7494.0  0.0532  0.0527  288.0  36.76
...         ...     ...     ...    ...    ...
2556.0   6806.0  0.0000  0.0000  296.0   0.00
2564.0  11534.0  0.0000  0.0000  296.0   0.00
2572.0   7317.0  0.0000  0.0000  296.0   0.00
2579.9  11412.0  0.0213  0.0212  288.0   0.34
2587.9   6100.0  0.0000  0.0000  296.0   0.00

[269 rows x 5 columns]

## Accessing specific wavelengths

As mentioned above, the first column represents the wavelength of the observation set. Some analysis may rely strictly on a specific wavelength range, which can be accessed like so. Notice how this indexing will not grab wavelengths just above the second number, e.g. 530.4 is not included below.

In [180]:
#Access a specific wavelength range for all observations
s.loc[518:530]

0                                     1                          \
          RAW    REF1    REF2     QA    RAD     RAW    REF1    REF2     QA   
518.4  5887.0  0.0487  0.0482  288.0  30.99  5955.0  0.0502  0.0497  288.0   
524.7  6375.0  0.0497  0.0492  288.0  33.66  6461.0  0.0513  0.0507  288.0   

              ...      36                                    37          \
         RAD  ...     RAW    REF1    REF2     QA    RAD     RAW    REF1   
518.4  31.96  ...  5881.0  0.0488  0.0483  288.0  30.78  5774.0  0.0468   
524.7  34.72  ...  6371.0  0.0498  0.0493  288.0  33.42  6233.0  0.0477   

                             
         REF2     QA    RAD  
518.4  0.0464  288.0  29.54  
524.7  0.0473  288.0  32.06  

[2 rows x 190 columns]

In [181]:
#Access a specific wavelength range for a single observation
s[0].loc[518:530]

RAW    REF1    REF2     QA    RAD
518.4  5887.0  0.0487  0.0482  288.0  30.99
524.7  6375.0  0.0497  0.0492  288.0  33.66

## Format Conversion
Finally, it is possible to convert the spectra object into any number of formats support by Pandas. For example, below, we convert the `.spc` file into CSV that can be opened and worked with in Excel.

In [201]:
s.to_csv('SP_2C_02_02358_S138_E3586.csv')

In [202]:
#Print the first 5 lines of the csv file that was saved to your working directory
#The first line will be the observation number, second will be the column labels, 
#and third will be the first row of data for wavelength 512.6
!head -n 3 SP_2C_02_02358_S138_E3586.csv 

,0,0,0,0,0,1,1,1,1,1,2,2,2,2,2,3,3,3,3,3,4,4,4,4,4,5,5,5,5,5,6,6,6,6,6,7,7,7,7,7,8,8,8,8,8,9,9,9,9,9,10,10,10,10,10,11,11,11,11,11,12,12,12,12,12,13,13,13,13,13,14,14,14,14,14,15,15,15,15,15,16,16,16,16,16,17,17,17,17,17,18,18,18,18,18,19,19,19,19,19,20,20,20,20,20,21,21,21,21,21,22,22,22,22,22,23,23,23,23,23,24,24,24,24,24,25,25,25,25,25,26,26,26,26,26,27,27,27,27,27,28,28,28,28,28,29,29,29,29,29,30,30,30,30,30,31,31,31,31,31,32,32,32,32,32,33,33,33,33,33,34,34,34,34,34,35,35,35,35,35,36,36,36,36,36,37,37,37,37,37
,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA,RAD,RAW,REF1,REF2,QA